### ISSUE TITLE
Add support for foreign keys referencing unique key columns in different order than in the unique key definition

### ISSUE THREAD

**Issue by absdjfh** [jOOQ/jOOQ#8553]
### Expected behavior and actual behavior:
For composite keys, when field order declared in the foreign key doesn't match field order declared in primary key, join on key from these tables returns no results due to incorrect join condition.

### Steps to reproduce the problem 
See https://github.com/absdjfh/jOOQ-mcve

Database Schema:
```
CREATE TABLE mcve.test1 (
  id1 INT NOT NULL,
  id2 INT NOT NULL,
  CONSTRAINT pk_test1 PRIMARY KEY (id1, id2)
);

CREATE TABLE mcve.test2 (
  id INT NOT NULL,
  test1_id1 INT NOT NULL,
  test1_id2 INT NOT NULL,
  CONSTRAINT pk_test2 PRIMARY KEY (id),
  CONSTRAINT fk_test2_test1 FOREIGN KEY (test1_id2, test1_id1) REFERENCES mcve.test1 (id2, id1)
);
```

Test:
```
ctx.insertInto(TEST1)
        .columns(TEST1.ID1, TEST1.ID2)
        .values(1, 2)
        .execute();

ctx.insertInto(TEST2)
        .columns(TEST2.ID, TEST2.TEST1_ID1, TEST2.TEST1_ID2)
        .values(1, 1, 2)
        .execute();

Result<Record> result = ctx.selectFrom(TEST1.innerJoin(TEST2).onKey(Keys.FK_TEST2_TEST1))
        .fetch();
assertEquals(1, result.size());
```

Generated SQL:
```
select "MCVE"."TEST1"."ID1", "MCVE"."TEST1"."ID2", "MCVE"."TEST2"."ID", "MCVE"."TEST2"."TEST1_ID1", "MCVE"."TEST2"."TEST1_ID2" 
from "MCVE"."TEST1" join "MCVE"."TEST2" 
on (
"MCVE"."TEST2"."TEST1_ID2" = "MCVE"."TEST1"."ID1" and 
"MCVE"."TEST2"."TEST1_ID1" = "MCVE"."TEST1"."ID2"
)
```

### Versions:

- jOOQ: 3.11.1
- Java: 8

**Comment by lukaseder** [comment]
Thank you very much for your report. Indeed, jOOQ's code generator and runtime assume that the foreign key column order must match the referenced primary key column order.

Indeed, the SQL standard allows for your syntax to be implemented, even if many RDBMS expect the column order to match that of a unique key specification (just like when you omit the column list and use `REFERENCES mcve.test1` only).

This could be fixed in 2 ways:

1. jOOQ-meta would ensure that reported foreign key columns always match the unique key column order, not the foreign key column order. This would fix your issue but produce the same wrong assumption about how foreign keys work in the runtime meta model, which causes issues when reproducing the DDL statement, e.g. through `DSLContext.ddl()`
2. We improve the runtime meta model and no longer match foreign key columns / primary key columns by index. The code generator would still have to match them by index, because we use column type information in type safe foreign key definitions when generating code.

Unfortunately, this doesn't look like an easy fix, which won't fit in 3.12 anymore (and is difficult to backport, once implemented). For the time being, if possible, I suggest matching keys by column index in your DDL and/or avoiding `onKey()`

**Comment by absdjfh** [comment]
Thanks, I've just recreated foreign key to match.
This issue was a little bit difficult to debug though. If it's easy to detect, I would suggest issuing a compile time warning (or even error) as a temporary solution.

**Comment by lukaseder** [comment]
I'm afraid I don't think it's very easy to detect as we currently don't have the meta information required to detect this. All we know in both the code generator and at runtime is the key column index.

If by accident, your key columns were of different types (e.g. `Integer` and `Long`), we could have an additional compile time check in generated code that would try to match key columns, but that would require new API as well, at which point we might as well tackle the actual fix.

**Comment by lukaseder** [comment]
Looking into this now. These are the steps:

- [ ] Add new runtime meta API
- [ ] Add interpreter and `org.jooq.Meta` support
- [ ] Add code generation support
- [ ] Update implementation for all dialects that support foreign keys, including:
  - [x] ~CUBRID~ (deprecated, won't fix)
  - [ ] DB2
  - [ ] Derby
  - [ ] Firebird
  - [ ] H2
  - [ ] HANA
  - [ ] HSQLDB
  - [ ] Informix
  - [ ] Ingres
  - [ ] MySQL
  - [ ] Oracle
  - [ ] PostgreSQL
  - [ ] Redshift
  - [ ] SQL Data Warehouse
  - [ ] SQLite
  - [ ] SQL Server
  - [ ] Sybase ASE
  - [ ] Sybase SQL Anywhere
  - [ ] Teradata
  - [ ] Vertica
  - [ ] XMLDatabase

**Comment by lukaseder** [comment]
Got a working implementation for `H2Database` (and by consequence `DDLDatabase` and others). Doesn't look so complicated after all, just a lot of effort of getting it right in all dialects.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.14
